Add docs for IdP-initiated SSO (#16093)
gravitational/teleport: docs/pages/access-controls/sso.mdx

      # Set as saml, oidc, or github
      type: saml|oidc|github
  ```

>   <Admonition type="note" title="IdP-initiated SSO">

Reviewer: - I think we should use a partial, since we include this Admonition in a number of locations

- I think a `Details` box would work better than an `Admonition` for the details about IdP-initiated SSO. It seems like this is an option that only a subset of users will use, and currently the Admonition takes up most of the space within this `TabItem`. 

- An example of using `allow_idp_initiated` in the auth connector configuration would be good since YAML is so sensitive to consistent whitespace.

Author: Amended, could you take another look now? @ptgott

    Enabling the `allow_idp_initiated` flag in SAML connectors allows users to

>     login to Teleport with one click from the dashboard provided by the IdP.

Reviewer: ```suggestion
    log in to Teleport with one click from the dashboard provided by the IdP.
```

    This feature is potentially unsafe and should be used with caution.

    Enabling IdP-initiated login comes with notable security risks such as:
    - Possibility of replay attacks on the SAML payload giving an attacker a secret web session

>     - Increased risk to man-in-the-middle based session hijacking and impersonation attacks

Reviewer: ```suggestion
    - Increased risk of session hijacking and impersonation attacks based on intercepting SAML communications
```
We would need to add another hyphen before "based", which looks awkward to me—suggested this rephrasing instead.

  # in the Web interface
  display: "Okta"

  # enables/disables idp-initiated saml login

>   allow_idp_initiated: false

Reviewer 2: Since false is the default, it's probably unlike to see this snippet in a connector, as people will only add `allow_idp_initiated` if they want to enable it. Should we change this example to `true` instead? (and probably add a note about the security implications)

Author: Hmm, I'm a bit wary of default-on since I'm afraid some may just copy-paste this or not read it thoroughly. I added this here for clarity so that astute readers may enable it if they so choose. Do you think it'd be worth it to change it to true? I'm 50/50 I feel like.

Reviewer 3: Could you pair a `true` value with the flag that still allows local user login, and then have them optionally remove that option later once the configuration is complete?

>   # The last segment of the URL must be identical to the connector metadata name.

Reviewer 2: Is this a new requirement, or it just wasn't documented before?

Author: This is a new requirement, existing connectors don't break, but IdP initiated logins will not work without it and will return an error code noting that the connector name isn't part of the ACS callback URI.

Reviewer 2: So does this new requirement only apply when enabling IdP-initiated login, or all the time going forward?

Author: It's only required with IdP-initiated login, but I'd say it's good practice to act like you should do it for simplicity's sake.

Reviewer 2: Okay, if we're introducing new constraints I think we need to mention it more than just a comment in the config YAML. In fact, a community user recently reported this error in slack and wasn't sure what do to about it.

Author: I have updated relevant guides to fulfill this constraint so you'll get this done correctly when you set up a new connector. Should we write an additional note/section about enabling this for existing connectors too?

Author: As discussed in our 1x1, I’ll be adding validation on connector upsert to ensure this cannot be violated. Docs should be good as-is.

  # in the Web interface
  display: "Okta"

  # enables/disables idp-initiated saml login
  allow_idp_initiated: false

>   # The last segment of the URL must be identical to the connector metadata name.

Reviewer 2: ```suggestion
  # The last segment of the URL must be identical to the connector metadata name
  # when IdP-initiated login is enabled.
```

</TabItem>
</Tabs>

<Admonition type="note" title="IdP-initiated SSO">
  Enabling the `allow_idp_initiated` flag in SAML connectors allows users to
  login to Teleport with one click from the dashboard provided by the IdP.

>   This comes with [multiple security considerations worth reviewing](https://github.com/gravitational/teleport/blob/master/rfd/0073-idp-initiated-login.md).

Reviewer 2: RFDs are intended for a very technical Teleport-developer audience, while product documentation should be more geared towards the average user.

Do you think we can highlight the major security concerns here without blowing up the page instead of linking to an RFD?

Author: I added this as an admonition for this reason. Although I can briefly list the security concerns in the form of the headlines used in the RFD, I don't believe that information alone would be useful to a Teleport cluster administrator and help them make an informed decision. I feel like as an admin I would want more detailed information to make a decision. I did consider writing a separate page for this but it ended up being very similar to the RFD itself.

Reviewer 2: Let's see what @ptgott thinks when he gets back from vacation. I'm still leaning towards _not_ linking to the RFD here, but I'm happy to be told I'm wrong.

Reviewer: First, I think we should move this Admonition into the appropriate `TabItem`s in the `Tabs` above so OIDC/GitHub users don't need to see irrelevant information.

If a reader needs to be familiar with a complex list of security considerations before enabling `allow_idp_initiated`, I don't think we can count on readers finding this `Admonition`, which is somewhat buried here. Instead, we can write a quick guide to IdP-initiated logins that spells out these considerations for Teleport's audience (i.e., paraphrasing but not linking to the RFD). We can then remove the mentions of `allow_idp_initiated` from elsewhere in the docs so readers don't accidentally enable it.

If a reader doesn't need to be familiar with all the considerations for using IdP-initiated logins, but only needs to know that it's generally unsafe, I think we can include one or two headline issues in this Admonition and remove the link. We'll also want to include similar Admonitions in our individual SSO guides, since a user might be arriving there from an organic search rather than from this guide.

Author: @ptgott I think the other solution works well. I've added per-tab admonitions and also per-page admonitions for relevant connector guide pages. Can you give this another look?
